### PR TITLE
Replace TimeLine LLVM profiling with the self profiler

### PR DESCRIPTION
**Pull request by wesleywiser**

### CODE REVIEW

File: src/librustc_codegen_ssa/base.rs [hunk 1]
```diff
@@ -676,15 +662,13 @@ pub fn codegen_crate<B: ExtraBackendMethods>(
...
-                                     CODEGEN_WORK_PACKAGE_KIND,
-                                     &format!("codegen {}", cgu.name()))
-                });
+                //TODO recover "codegen {}" cgu.name() info
+                tcx.sess.profiler(|p| p.start_activity(ProfileCategory::Codegen));
                 let start_time = Instant::now();
                 let stats = backend.compile_codegen_unit(tcx, *cgu.name());
                 all_stats.extend(stats);
                 total_codegen_time += start_time.elapsed();
+                tcx.sess.profiler(|p| p.start_activity(ProfileCategory::Codegen));
```
Comment (michaelwoerister) [inline src/librustc_codegen_ssa/base.rs]: This should probably be `end_activity`.

**Comment by michaelwoerister** [comment]
Yes, I think this is going in the right direction. However, I'd hold off on doing "instant" events at the moment and only use events with a distinct beginning and end. That is, basically, add a label to `GenericActivityStart` and `GenericActivityEnd` and remove `GenericEvent`. 

Let me do another pass over the existing `TimeLine` events and mark them as either "remove" or "start-end"...

File: src/librustc_codegen_llvm/back/lto.rs [hunk 2]
```diff
@@ -69,7 +67,7 @@ fn prepare_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
         .iter()
         .filter_map(symbol_filter)
         .collect::<Vec<CString>>();
-    timeline.record("whitelist");
+    cgcx.profile(|p| p.generic_event("whitelist"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: `start-end` event extending to the rest of the function.

File: src/librustc_codegen_llvm/back/lto.rs [hunk 3]
```diff
@@ -122,7 +120,7 @@ fn prepare_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
                 let bc = SerializedModule::FromRlib(bc);
                 upstream_modules.push((bc, CString::new(id).unwrap()));
             }
-            timeline.record(&format!("load: {}", path.display()));
+            cgcx.profile(|p| p.generic_event(format!("load: {}", path.display())));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: `start-end` event spanning the loop body of `for &(cnum, ref path) in cgcx.each_linked_rlib_for_lto.iter() { ...`

File: src/librustc_codegen_llvm/back/lto.rs [hunk 4]
```diff
@@ -304,7 +297,7 @@ fn fat_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
                     write::llvm_err(&diag_handler, &msg)
                 })
             })?;
-            timeline.record(&format!("link {:?}", name));
+            cgcx.profile(|p| p.generic_event(format!("link {:?}", name)));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: can be removed for now

File: src/librustc_codegen_llvm/back/lto.rs [hunk 5]
```diff
@@ -326,7 +319,7 @@ fn fat_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
             }
             save_temp_bitcode(&cgcx, &module, "lto.after-nounwind");
         }
-        timeline.record("passes");
+        cgcx.profile(|p| p.generic_event("passes"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: can be removed for now

File: src/librustc_codegen_llvm/back/lto.rs [hunk 6]
```diff
@@ -423,7 +415,7 @@ fn thin_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
             });
             thin_buffers.push(buffer);
             module_names.push(cname);
-            timeline.record(&name);
+            cgcx.profile(|p| p.generic_event(name));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: can be removed for now

File: src/librustc_codegen_llvm/back/lto.rs [hunk 7]
```diff
@@ -476,7 +468,7 @@ fn thin_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
         })?;
 
         info!("thin LTO data created");
-        timeline.record("data");
+        cgcx.profile(|p| p.generic_event("data"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: can be removed for now

File: src/librustc_codegen_llvm/back/lto.rs [hunk 8]
```diff
@@ -487,7 +479,7 @@ fn thin_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
             ThinLTOImports::default()
         };
         info!("thin LTO import map loaded");
-        timeline.record("import-map-loaded");
+        cgcx.profile(|p| p.generic_event("import-map-loaded"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: can be removed for now

File: src/librustc_codegen_llvm/back/lto.rs [hunk 9]
```diff
@@ -741,7 +732,7 @@ pub unsafe fn optimize_thin_module(
         if cgcx.no_landing_pads {
             llvm::LLVMRustMarkAllFunctionsNounwind(llmod);
             save_temp_bitcode(&cgcx, &module, "thin-lto-after-nounwind");
-            timeline.record("nounwind");
+            cgcx.profile(|p| p.generic_event("nounwind"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: `start-end` activity spanning the body of the `if`

File: src/librustc_codegen_llvm/back/lto.rs [hunk 10]
```diff
@@ -808,7 +799,7 @@ pub unsafe fn optimize_thin_module(
         // fixed in LLVM.
         llvm::LLVMRustThinLTOPatchDICompileUnit(llmod, cu1);
         save_temp_bitcode(cgcx, &module, "thin-lto-after-patch");
-        timeline.record("patch");
+        cgcx.profile(|p| p.generic_event("patch"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: The above events can all be removed for now.

File: src/librustc_codegen_llvm/back/lto.rs [hunk 11]
```diff
@@ -819,7 +810,7 @@ pub unsafe fn optimize_thin_module(
         let config = cgcx.config(module.kind);
         run_pass_manager(cgcx, &module, config, true);
         save_temp_bitcode(cgcx, &module, "thin-lto-after-pm");
-        timeline.record("thin-done");
+        cgcx.profile(|p| p.generic_event("thin-done"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: This one too...

File: src/librustc_codegen_llvm/back/write.rs [hunk 12]
```diff
@@ -422,7 +420,7 @@ pub(crate) unsafe fn optimize(cgcx: &CodegenContext<LlvmCodegenBackend>,
                  || {
             llvm::LLVMRustRunFunctionPassManager(fpm, llmod)
         });
-        timeline.record("fpm");
+        cgcx.profile(|p| p.generic_event("fpm"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: `start-end` activity with the same span as the `time_ext` call.

File: src/librustc_codegen_llvm/back/write.rs [hunk 13]
```diff
@@ -422,7 +420,7 @@ pub(crate) unsafe fn optimize(cgcx: &CodegenContext<LlvmCodegenBackend>,
...
         time_ext(config.time_passes,
                  None,
                  &format!("llvm module passes [{}]", module_name.unwrap()),
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: This `time_ext` call should also have `start-end` activity called `module passes` or something like that.

File: src/librustc_codegen_llvm/back/write.rs (hunk 12 above)
Comment (michaelwoerister) [reply src/librustc_codegen_llvm/back/write.rs]: renaming to `function passes` would be good too.

File: src/librustc_codegen_llvm/back/write.rs [hunk 14]
```diff
@@ -440,11 +438,10 @@ pub(crate) unsafe fn optimize(cgcx: &CodegenContext<LlvmCodegenBackend>,
 pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
                   diag_handler: &Handler,
                   module: ModuleCodegen<ModuleLlvm>,
-                  config: &ModuleConfig,
-                  timeline: &mut Timeline)
+                  config: &ModuleConfig)
     -> Result<CompiledModule, FatalError>
 {
-    timeline.record("codegen");
+    cgcx.profile(|p| p.generic_event("codegen"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: `start-end` activity that spans the entire function body of `fn codegen()`. The sub events in here can also be converted into `start-end` events or removed for now.

File: src/librustc_codegen_ssa/back/lto.rs [hunk 15]
```diff
@@ -67,19 +66,18 @@ impl<B: WriteBackendMethods> LtoModuleCodegen<B> {
...
-        timeline: &mut Timeline
     ) -> Result<ModuleCodegen<B::Module>, FatalError> {
         match *self {
             LtoModuleCodegen::Fat { ref mut module, .. } => {
                 let module = module.take().unwrap();
                 {
                     let config = cgcx.config(module.kind);
                     B::run_lto_pass_manager(cgcx, &module, config, false);
-                    timeline.record("fat-done");
+                    cgcx.profile(|p| p.generic_event("fat-done"));
```
Comment (michaelwoerister) [inline src/librustc_codegen_ssa/back/lto.rs]: remove for now...

**Comment by wesleywiser** [comment]
Thanks for all the feedback @michaelwoerister!

File: src/librustc/session/mod.rs [hunk 16]
```diff
@@ -1103,11 +1099,12 @@ pub fn build_session_with_source_map(
         },
     );
 
-    build_session_(sopts, local_crate_source_file, diagnostic_handler, source_map)
+    build_session_(sopts, self_profiler, local_crate_source_file, diagnostic_handler, source_map)
 }
 
 pub fn build_session_(
     sopts: config::Options,
+    self_profiler: Option<Arc<Mutex<SelfProfiler>>>,
```
Comment (Zoxc) [inline src/librustc/session/mod.rs]: Why is `self_profiler` passed an argument here?

Comment (wesleywiser) [reply src/librustc/session/mod.rs]: We need it to pass on line 1189: https://github.com/rust-lang/rust/pull/58488/files#diff-1c226b200b9385d77428f0b4418366e4R1189

Comment (Zoxc) [reply src/librustc/session/mod.rs]: Why isn't it created there like it was before?

Comment (wesleywiser) [reply src/librustc/session/mod.rs]: We wanted to create it as early as possible so we didn't lose time at the very beginning of the process. So we're creating it here https://github.com/rust-lang/rust/pull/58488/commits/7cc697641d7050682f9f1c768529d3839276269a#diff-707a0eda6b2f1a0537abc3d23133748cR466

Comment (michaelwoerister) [reply src/librustc/session/mod.rs]: Yes, what was the number you had, @wesleywiser? Something like 30% of the time for compiling small crates was unaccounted for...

Comment (wesleywiser) [reply src/librustc/session/mod.rs]: I think it was closer to 10% for `hello_world.rs`. In practice, for a reasonably sized crate like `regex` or something, it was < 2%.

**Comment by Dylan-DPC** [comment]
ping from triage @wesleywiser any updates?

**Comment by wesleywiser** [comment]
Rebased & comments addressed. Should be ready for review again.

File: src/librustc_codegen_llvm/back/lto.rs [hunk 17]
```diff
@@ -68,7 +67,7 @@ fn prepare_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
         .iter()
         .filter_map(symbol_filter)
         .collect::<Vec<CString>>();
-    timeline.record("whitelist");
+    let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "whitelist");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: "generate_symbol_white_list_for_thinlto"

File: src/librustc_codegen_llvm/back/lto.rs [hunk 18]
```diff
@@ -738,9 +726,9 @@ pub unsafe fn optimize_thin_module(
         // Like with "fat" LTO, get some better optimizations if landing pads
         // are disabled by removing all landing pads.
         if cgcx.no_landing_pads {
+            let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "nounwind");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/lto.rs]: "LLVM_remove_landing_pads"

File: src/librustc_codegen_llvm/back/write.rs [hunk 19]
```diff
@@ -415,19 +414,24 @@ pub(crate) unsafe fn optimize(cgcx: &CodegenContext<LlvmCodegenBackend>,
...
-        });
-        timeline.record("fpm");
-        time_ext(config.time_passes,
-                 None,
-                 &format!("llvm module passes [{}]", module_name.unwrap()),
-                 || {
-            llvm::LLVMRunPassManager(mpm, llmod)
-        });
+        {
+            let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "function passes");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_function_passes"

File: src/librustc_codegen_llvm/back/write.rs [hunk 20]
```diff
@@ -415,19 +414,24 @@ pub(crate) unsafe fn optimize(cgcx: &CodegenContext<LlvmCodegenBackend>,
...
+            time_ext(config.time_passes,
+                        None,
+                        &format!("llvm function passes [{}]", module_name.unwrap()),
+                        || {
+                llvm::LLVMRustRunFunctionPassManager(fpm, llmod)
+            });
+        }
+        {
+            let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "module passes");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_module_passes"

File: src/librustc_codegen_llvm/back/write.rs [hunk 21]
```diff
@@ -494,29 +497,29 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
 
 
         if write_bc || config.emit_bc_compressed || config.embed_bitcode {
+            let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "make-bc");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_make_bitcode"

File: src/librustc_codegen_llvm/back/write.rs [hunk 22]
```diff
@@ -494,29 +497,29 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
...
             let thin = ThinBuffer::new(llmod);
             let data = thin.data();
-            timeline.record("make-bc");
 
             if write_bc {
+                let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "write-bc");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_emit_bitcode"

File: src/librustc_codegen_llvm/back/write.rs [hunk 23]
```diff
@@ -494,29 +497,29 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
...
                 if let Err(e) = fs::write(&bc_out, data) {
                     diag_handler.err(&format!("failed to write bytecode: {}", e));
                 }
-                timeline.record("write-bc");
             }
 
             if config.embed_bitcode {
+                let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "embed-bc");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_embed_bitcode"

File: src/librustc_codegen_llvm/back/write.rs [hunk 24]
```diff
@@ -494,29 +497,29 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
...
                 embed_bitcode(cgcx, llcx, llmod, Some(data));
-                timeline.record("embed-bc");
             }
 
             if config.emit_bc_compressed {
+                let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "compress-bc");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_compress_bitcode"

File: src/librustc_codegen_llvm/back/write.rs [hunk 25]
```diff
@@ -525,6 +528,7 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
         time_ext(config.time_passes, None, &format!("codegen passes [{}]", module_name.unwrap()),
             || -> Result<(), FatalError> {
             if config.emit_ir {
+                let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "ir");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_emit_ir"

File: src/librustc_codegen_llvm/back/write.rs [hunk 26]
```diff
@@ -563,10 +567,10 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
                     llvm::LLVMRustPrintModule(cpm, llmod, out.as_ptr(), demangle_callback);
                     llvm::LLVMDisposePassManager(cpm);
                 });
-                timeline.record("ir");
             }
 
             if config.emit_asm || asm_to_obj {
+                let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "asm");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_emit_asm"

File: src/librustc_codegen_llvm/back/write.rs [hunk 27]
```diff
@@ -581,19 +585,18 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
                     write_output_file(diag_handler, tm, cpm, llmod, &path,
                                       llvm::FileType::AssemblyFile)
                 })?;
-                timeline.record("asm");
             }
 
             if write_obj {
+                let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "obj");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_emit_obj"

File: src/librustc_codegen_llvm/back/write.rs [hunk 28]
```diff
@@ -581,19 +585,18 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<LlvmCodegenBackend>,
...
                 with_codegen(tm, llmod, config.no_builtins, |cpm| {
                     write_output_file(diag_handler, tm, cpm, llmod, &obj_out,
                                       llvm::FileType::ObjectFile)
                 })?;
-                timeline.record("obj");
             } else if asm_to_obj {
+                let _timer = cgcx.profile_activity(ProfileCategory::Codegen, "asm_to_obj");
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/back/write.rs]: "LLVM_asm_to_obj"

File: src/librustc_codegen_llvm/lib.rs [hunk 29]
```diff
@@ -335,12 +328,12 @@ impl CodegenBackend for LlvmCodegenBackend {
 
         // Run the linker on any artifacts that resulted from the LLVM run.
         // This should produce either a finished executable or library.
-        sess.profiler(|p| p.start_activity(ProfileCategory::Linking));
+        sess.profiler(|p| p.start_activity(ProfileCategory::Linking, ""));
```
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/lib.rs]: "link_crate"

File: src/librustc_codegen_ssa/back/write.rs [hunk 30]
```diff
@@ -285,37 +324,36 @@ fn generate_lto_work<B: ExtraBackendMethods>(
     needs_thin_lto: Vec<(String, B::ThinBuffer)>,
     import_only_modules: Vec<(SerializedModule<B::ModuleBuffer>, WorkProduct)>
 ) -> Vec<(WorkItem<B>, u64)> {
-    let mut timeline = cgcx.time_graph.as_ref().map(|tg| {
-        tg.start(CODEGEN_WORKER_TIMELINE,
-                 CODEGEN_WORK_PACKAGE_KIND,
-                 "generate lto")
-    }).unwrap_or(Timeline::noop());
+    cgcx.profile(|p| p.start_activity(ProfileCategory::Linking, ""));
```
Comment (michaelwoerister) [inline src/librustc_codegen_ssa/back/write.rs]: "codegen_run_lto"

File: src/librustc_driver/driver.rs [hunk 31]
```diff
@@ -647,15 +647,15 @@ pub fn phase_1_parse_input<'a>(
         profile::begin(sess);
     }
 
-    sess.profiler(|p| p.start_activity(ProfileCategory::Parsing));
+    sess.profiler(|p| p.start_activity(ProfileCategory::Parsing, ""));
```
Comment (michaelwoerister) [inline src/librustc_driver/driver.rs]: "parsing"

File: src/librustc_driver/driver.rs [hunk 32]
```diff
@@ -916,7 +916,7 @@ where
     syntax_ext::register_builtins(&mut resolver, syntax_exts);
 
     // Expand all macros
-    sess.profiler(|p| p.start_activity(ProfileCategory::Expansion));
+    sess.profiler(|p| p.start_activity(ProfileCategory::Expansion, ""));
```
Comment (michaelwoerister) [inline src/librustc_driver/driver.rs]: "macro expansion"

File: src/librustc_driver/driver.rs [hunk 33]
```diff
@@ -1233,9 +1233,9 @@ pub fn phase_4_codegen<'a, 'tcx>(
         ::rustc::middle::dependency_format::calculate(tcx)
     });
 
-    tcx.sess.profiler(|p| p.start_activity(ProfileCategory::Codegen));
+    tcx.sess.profiler(|p| p.start_activity(ProfileCategory::Codegen, ""));
```
Comment (michaelwoerister) [inline src/librustc_driver/driver.rs]: "codegen crate"

File: src/librustc_codegen_ssa/base.rs [hunk 34]
```diff
@@ -545,7 +539,7 @@ pub fn codegen_crate<B: ExtraBackendMethods>(
     let cgu_name_builder = &mut CodegenUnitNameBuilder::new(tcx);
 
     // Codegen the metadata.
-    tcx.sess.profiler(|p| p.start_activity(ProfileCategory::Codegen));
+    tcx.sess.profiler(|p| p.start_activity(ProfileCategory::Codegen, ""));
```
Comment (michaelwoerister) [inline src/librustc_codegen_ssa/base.rs]: "codegen crate metadata"

File: src/librustc_typeck/lib.rs [hunk 35]
```diff
@@ -317,7 +317,7 @@ pub fn provide(providers: &mut Providers<'_>) {
 pub fn check_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>)
                              -> Result<(), ErrorReported>
 {
-    tcx.sess.profiler(|p| p.start_activity(ProfileCategory::TypeChecking));
+    tcx.sess.profiler(|p| p.start_activity(ProfileCategory::TypeChecking, ""));
```
Comment (michaelwoerister) [inline src/librustc_typeck/lib.rs]: "type-check crate"

**Review by michaelwoerister — COMMENTED** [review]
Looks good!

I think we shouldn't leave events unnamed since the "category" will be going away soon. I added suggestions on what to name things below.

**Comment by wesleywiser** [comment]
Fixed. Thanks!

**Comment by michaelwoerister** [comment]
Thanks, @wesleywiser!

@bors r+

**Comment by wesleywiser** [comment]
Spurious network error:

> warning: spurious network error (2 tries remaining): failed to get 200 response from `https://static.crates.io/crates/dtoa/dtoa-0.4.1.crate`, got 504

> warning: spurious network error (1 tries remaining): failed to get 200 response from `https://static.crates.io/crates/dtoa/dtoa-0.4.1.crate`, got 504

> error: failed to download from `https://crates.io/api/v1/crates/dtoa/0.4.1/download`

@bors retry

**Comment by kennytm** [comment]
@bors retry

4 hour timeout, a job didn't start at all.

**Comment by wesleywiser** [comment]
Looks spurious to me

@bors retry

**Pull request merged**